storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java

```diff
@@ -50,10 +61,28 @@
     volatile int[] choices;
     private volatile int[] prepareChoices;
     private AtomicInteger current;
+    private AtomicReference<Scope> currentScopeRef;
```

Comment:
I don't think this needs to be atomic.  Only `chooseTasks` needs to be thread safe.  All the others are called from a single thread, and we can put a lock around if we really want to.

Reply:
I think you are right